packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -41,13 +57,24 @@ export type SelectorEventHandler<SelectorReturnValue> = (
 
 export type ActionConstraint = {
   type: string;
-  handler: (...args: any) => unknown;
+  handler: (...args: never[]) => unknown;
```

Comment:
Highlighting to request closer review.

Reply:
Similar; this doesn't seem right. Does `unknown[]` work instead of `never[]`?

Reply:
This is a really fun one! So just in empirical terms, it's already clear that `never[]` is correct, both here and in `ExtractActionResponse`. Changing either to `unknown[]` (or even `[never]`) triggers a whole lot of errors all over the codebase, basically everywhere the `Action` type is used. 

But this is also logically sound, and the reason for this is interesting. 

Here we have `Action['handler'] extends ActionConstraint['handler']`, and in general `Subtype extends Supertype`. So we need `ActionConstraint['handler']` to be a function that is a supertype of all functions. 

To achieve this, the return type needs to be `unknown` (universal supertype), and the parameters need to be `never` (universal subtype). 
This is because while `number | string` is a supertype of `number`, `(x: number | string) => void` is a subtype of `(x: number) => void`.

- There's an example of this in the [TypeScript docs](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types)
- And here's a more detailed [reddit comment](https://www.reddit.com/r/typescript/comments/muyl55/comment/gv9ndij/?utm_source=share&utm_medium=web2x&context=3) on this topic.

Reply:
Regarding `never[]`, see: https://github.com/MetaMask/core/pull/1890#issuecomment-1783050539